src/liballoc/sync.rs

```diff
@@ -322,11 +323,43 @@ impl<T> Arc<T> {
         let x: Box<_> = box ArcInner {
             strong: atomic::AtomicUsize::new(1),
             weak: atomic::AtomicUsize::new(1),
-            data,
+            data: UnsafeCell::new(data),
         };
         Self::from_inner(Box::leak(x).into())
     }
 
+    /// Constructs a new `Arc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function returns will result
+    /// in a `None` value.
+    #[inline]
+    #[unstable(feature = "arc_new_with", issue = "none")]
+    pub fn new_with(data_fn: impl FnOnce(Weak<T>) -> T) -> Arc<T> {
+        // Start the weak pointer count at 2:
+        // - One for the pointer we'll pass into the constructor.
+        // - One for ourselves to ensure the memory is not freed.
+        let uninit_ptr: NonNull<_> = Box::leak(box ArcInner {
+            strong: atomic::AtomicUsize::new(0),
+            weak: atomic::AtomicUsize::new(2),
+            data: UnsafeCell::new(mem::MaybeUninit::<T>::uninit()),
+        })
+        .into();
+        let ptr: NonNull<ArcInner<T>> = uninit_ptr.cast();
+
+        let weak1 = Weak { ptr };
+        let weak2 = Weak { ptr };
```

Comment:
`weak2`'s only use is a as a guard to ensure the box is dropped on unwinding (This seems non-obvious enough to be a comment, maybe).